README.md
*** IMPORTANT NOTICES ***

If you are looking for an alternative of [actoins/setup-ruby](https://github.com/actions/setup-ruby) and you have no special reason for building Ruby by yourself, I recommend you to try [eregon/use-ruby-action](https://github.com/eregon/use-ruby-action).  This action, which has been developed by a Ruby committer, does installing pre-compiled Rubies built at [eregon/ruby-install-builder](https://github.com/eregon/ruby-install-builder). It is much faster than this action and I believe it is what most Ruby developers want to have.

# setup-rbenv

Setup Ruby by [rbenv](https://github.com/rbenv/rbenv) on Ubuntu 18.04 / 16.04 virtual environments, not using tool-cache provided by GitHub to catch up the latest versions of Ruby.

# Usage

You need to initialize rbenv by dispatching `eval "$(rbenv init -)"` in each step before you run Ruby.

```yaml
steps:
- uses: actions/checkout@master
- uses: masa-iwasaki/setup-rbenv@v1
- name: Install Ruby
  run: |
    eval "$(rbenv init -)"
    rbenv install -s `cat .ruby-version` # or specify the version you want

- name: Run bundler
  run: |
    eval "$(rbenv init -)"
    bundle install --path vendor/bundle
```

## Cache

By using [actions/cache](https://github.com/actions/cache), you can keep your rubies in your cache.

```yaml
steps:
- uses: actions/checkout@master
- uses: masa-iwasaki/setup-rbenv@v1

- name: Cache RBENV_ROOT
  uses: actions/cache@v1
  id: cache_rbenv
  with:
    path: /home/runner/.rbenv
    key: ${{ runner.os }}-rbenv-${{ hashFiles('**/.ruby-version') }}
    restore-keys: |
      {{ runner.os }}-rbenv-

- name: Install Ruby
  run: |
    eval "$(rbenv init -)"
    rbenv install -s `cat .ruby-version`
```

You don't need to check cache hit because `-s` option for `rbenv install` does the job. You need to be aware of [cache limits](https://github.com/actions/cache#cache-limits) if you want to keep multiple rubies.

# License

The scripts and documentation in this project are released under the [MIT License](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2020-01-06

- Use `--depth 1` option for `git clone` to reduce disk usage under RBENV_ROOT ([#5](https://github.com/masa-iwasaki/setup-rbenv/pull/5))

## [1.1.0] - 2020-01-01

### Added

- Support Ubuntu 16.04 ([#3](https://github.com/masa-iwasaki/setup-rbenv/pull/3), [#4](https://github.com/masa-iwasaki/setup-rbenv/pull/4))
- Fix README ([#2](https://github.com/masa-iwasaki/setup-rbenv/pull/2))

## [1.0.0] - 2019-12-10

The first release :tada:

__tests__/main.test.ts
// import * as installer from '../src/installer'

test("TODO: Add tests", () => {
    expect(1).toBeGreaterThan(0)
});

lib/main.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const core = __importStar(require("@actions/core"));
const installer = __importStar(require("./installer"));
function run() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            if (process.platform !== 'linux') {
                core.error(`Not a supported platform. Only Linux is supported.`);
            }
            let options = {
                rbenvRoot: "/home/runner/.rbenv",
                rbenvRootOwner: "runner"
            };
            let rbenvRoot = core.getInput('rbenv_root');
            let rbenvRootOwner = core.getInput('rbenv_root_owner');
            if (!!rbenvRoot) {
                options.rbenvRoot = rbenvRoot;
            }
            if (!!rbenvRootOwner) {
                options.rbenvRootOwner = rbenvRootOwner;
            }
            yield installer.intallRbenv(options);
            yield installer.installRubyBuild(options);
            core.exportVariable('RBENV_ROOT', options.rbenvRoot);
            core.addPath(`${options.rbenvRoot}/bin`);
        }
        catch (error) {
            core.setFailed(error.message);
        }
    });
}
run();


src/main.ts
import * as core from '@actions/core';
import * as installer from './installer';

async function run() {
  try {

    if (process.platform !== 'linux') {
      core.error(`Not a supported platform. Only Linux is supported.`);
    }

    let options: installer.RbenvOptions = {
      rbenvRoot: "/home/runner/.rbenv",
      rbenvRootOwner: "runner"
    }

    let rbenvRoot = core.getInput('rbenv_root');
    let rbenvRootOwner = core.getInput('rbenv_root_owner');

    if (!!rbenvRoot) {
      options.rbenvRoot = rbenvRoot;
    }

    if (!!rbenvRootOwner) {
      options.rbenvRootOwner = rbenvRootOwner;
    }

    await installer.intallRbenv(options);
    await installer.installRubyBuild(options);

    core.exportVariable('RBENV_ROOT', options.rbenvRoot);
    core.addPath(`${options.rbenvRoot}/bin`);
  } catch (error) {
    core.setFailed(error.message);
  }
}

run();


src/installer.ts
import * as exec from '@actions/exec';
import * as fs from 'fs'

export async function intallRbenv(options: RbenvOptions) {
  await exec.exec('sudo', ['git', 'clone', '--depth', '1', 'https://github.com/rbenv/rbenv.git', options.rbenvRoot]);
}

export interface RbenvOptions {
  rbenvRoot: string;
  rbenvRootOwner: string;
};

export async function installRubyBuild(options: RbenvOptions) {
  // from https://github.com/rbenv/ruby-build/wiki
  const packages = [
    "libreadline6-dev",
    "zlib1g-dev",
    "libncurses5-dev",
    "libffi-dev",
    "libgdbm-dev"
  ];

  const osRelease = fs.readFileSync("/etc/os-release", {encoding: "utf-8"});
  const matched = osRelease.match(/VERSION_ID="([0-9.]+)"/);
  const ubuntuVersion = matched ? matched[1] : "0";

  if (parseFloat(ubuntuVersion) >= 18.04 ){
    // Ubuntu 18.04+ (bionic)
    packages.push("libgdbm5");
  } else {
    // Ubuntu 16.04 (xenial)
    packages.push("libgdbm3");
  }

  const rubyBuildInstallPath = `${options.rbenvRoot}/plugins/ruby-build`;

  await exec.exec('sudo', ['apt-get', 'update']);
  await exec.exec('sudo', ['apt-get', 'install', '-y', ...packages]);

  await exec.exec('sudo', ['git', 'clone', '--depth', '1', 'https://github.com/rbenv/ruby-build.git', rubyBuildInstallPath]);

  await exec.exec('sudo', ['chown', '-R', options.rbenvRootOwner, options.rbenvRoot])
};


lib/installer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const exec = __importStar(require("@actions/exec"));
const fs = __importStar(require("fs"));
function intallRbenv(options) {
    return __awaiter(this, void 0, void 0, function* () {
        yield exec.exec('sudo', ['git', 'clone', '--depth', '1', 'https://github.com/rbenv/rbenv.git', options.rbenvRoot]);
    });
}
exports.intallRbenv = intallRbenv;
;
function installRubyBuild(options) {
    return __awaiter(this, void 0, void 0, function* () {
        // from https://github.com/rbenv/ruby-build/wiki
        const packages = [
            "libreadline6-dev",
            "zlib1g-dev",
            "libncurses5-dev",
            "libffi-dev",
            "libgdbm-dev"
        ];
        const osRelease = fs.readFileSync("/etc/os-release", { encoding: "utf-8" });
        const matched = osRelease.match(/VERSION_ID="([0-9.]+)"/);
        const ubuntuVersion = matched ? matched[1] : "0";
        if (parseFloat(ubuntuVersion) >= 18.04) {
            // Ubuntu 18.04+ (bionic)
            packages.push("libgdbm5");
        }
        else {
            // Ubuntu 16.04 (xenial)
            packages.push("libgdbm3");
        }
        const rubyBuildInstallPath = `${options.rbenvRoot}/plugins/ruby-build`;
        yield exec.exec('sudo', ['apt-get', 'update']);
        yield exec.exec('sudo', ['apt-get', 'install', '-y', ...packages]);
        yield exec.exec('sudo', ['git', 'clone', '--depth', '1', 'https://github.com/rbenv/ruby-build.git', rubyBuildInstallPath]);
        yield exec.exec('sudo', ['chown', '-R', options.rbenvRootOwner, options.rbenvRoot]);
    });
}
exports.installRubyBuild = installRubyBuild;
;
